src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableUpcallHandler.java

```diff
@@ -28,115 +28,287 @@
+                    methodType(Object.class, Object[].class, MethodHandle.class, Map.class, Map.class,
+                            CallingSequence.class, long.class));
+        } catch (ReflectiveOperationException e) {
+            throw new InternalError(e);
+        }
+    }
+
+    public static UpcallHandler make(ABIDescriptor abi, MethodHandle target, CallingSequence callingSequence) {
+        Binding.VMLoad[] argMoves = argMoveBindings(callingSequence);
+        Binding.VMStore[] retMoves = retMoveBindings(callingSequence);
+
+        boolean isSimple = !(retMoves.length > 1);
+
+        Class<?> llReturn = !isSimple
+            ? Object[].class
+            : retMoves.length == 1
+                ? retMoves[0].type()
+                : void.class;
+        Class<?>[] llParams = Arrays.stream(argMoves).map(Binding.Move::type).toArray(Class<?>[]::new);
+        MethodType llType = MethodType.methodType(llReturn, llParams);
+
+        MethodHandle doBindings;
+        long bufferCopySize = SharedUtils.bufferCopySize(callingSequence);
+        if (USE_SPEC && isSimple) {
+            doBindings = specializedBindingHandle(target, callingSequence, llReturn, bufferCopySize);
+            assert doBindings.type() == llType;
+        } else {
+            Map<VMStorage, Integer> argIndices = SharedUtils.indexMap(argMoves);
+            Map<VMStorage, Integer> retIndices = SharedUtils.indexMap(retMoves);
+            target = target.asSpreader(Object[].class, callingSequence.methodType().parameterCount());
+            doBindings = insertArguments(MH_invokeInterpBindings, 1, target, argIndices, retIndices, callingSequence,
+                    bufferCopySize);
+            doBindings = doBindings.asCollector(Object[].class, llType.parameterCount());
+            doBindings = doBindings.asType(llType);
+        }
+
+        long entryPoint;
+        boolean usesStackArgs = argMoveBindingsStream(callingSequence)
+                .map(Binding.VMLoad::storage)
+                .anyMatch(s -> abi.arch.isStackType(s.type()));
+        if (USE_INTRINSICS && isSimple && !usesStackArgs && supportsOptimizedUpcalls()) {
+            checkPrimitive(doBindings.type());
+            JLI.ensureCustomized(doBindings);
+            VMStorage[] args = Arrays.stream(argMoves).map(Binding.Move::storage).toArray(VMStorage[]::new);
+            VMStorage[] rets = Arrays.stream(retMoves).map(Binding.Move::storage).toArray(VMStorage[]::new);
+            CallRegs conv = new CallRegs(args, rets);
+            entryPoint = allocateOptimizedUpcallStub(doBindings, abi, conv);
+        } else {
+            BufferLayout layout = BufferLayout.of(abi);
+            MethodHandle doBindingsErased = doBindings.asSpreader(Object[].class, doBindings.type().parameterCount());
+            MethodHandle invokeMoves = insertArguments(MH_invokeMoves, 1, doBindingsErased, argMoves, retMoves, abi, layout);
+            InterpretedHandler handler = new InterpretedHandler(invokeMoves);
```

Comment:
In principle, this could be just replaced by the method handle no? Then we could have a static entry point in this class which can be used by the VM to "call" the method handle with given args. E.g. do we need the `InterpreterHandler` wrapper?

Reply:
Probably not. I think I had more fields in there before, so it's a leftover.